src/main/scala/it/pagopa/interop/backendforfrontend/api/impl/EServicesApiServiceImpl.scala

```diff
@@ -205,37 +206,32 @@ final case class EServicesApiServiceImpl(
     }
   }
 
-  private def getProducerEServicesIds(producerId: UUID, consumersIds: List[String])(implicit
+  private def getProducerEServicesIds(producerId: UUID, consumersUUIDs: List[UUID])(implicit
     contexts: Seq[(String, String)]
-  ): Future[List[UUID]] = {
-    for {
-      consumersIdsUUID <- Future.traverse(consumersIds)(_.toFutureUUID)
-
-      agreements <- Future
-        .traverse(consumersIdsUUID)(consumerId => {
-          loop(List())(0) { offset =>
-            agreementProcessService.getAgreements(
-              consumersIds = Seq(consumerId),
-              producersIds = Seq(producerId),
-              states = Seq(AgreementProcess.AgreementState.ACTIVE, AgreementProcess.AgreementState.SUSPENDED),
-              limit = 50,
-              offset = offset
-            )
-          }
-        })
-        .map(_.flatten)
-    } yield {
-      agreements.map(_.eserviceId).distinct
-    }
-  }
-  // It's stack safe since every submission to an Execution context resets the stack, creating a trampoline effect
-  def loop(
-    acc: List[AgreementProcess.Agreement]
-  )(offset: Int)(f: Int => Future[AgreementProcess.Agreements]): Future[List[AgreementProcess.Agreement]] = {
-    f(offset).flatMap(res =>
-      if (res.totalCount < offset + 50) Future.successful(acc ++ res.results)
-      else loop(acc ++ res.results)(offset + 50)(f)
-    )
+  ): Future[List[UUID]] = getAllAgreements(producerId, consumersUUIDs)
+    .map(_.map(_.eserviceId).distinct)
+
+  private def getAllAgreements(producerId: UUID, consumerUUIDs: List[UUID])(implicit
+    contexts: Seq[(String, String)]
+  ): Future[List[AgreementProcess.Agreement]] = {
+
+    def getAgreementsFrom(offset: Int): Future[List[AgreementProcess.Agreement]] =
+      agreementProcessService
+        .getAgreements(
+          producersIds = producerId :: Nil,
+          consumersIds = consumerUUIDs,
+          states = Seq(AgreementProcess.AgreementState.ACTIVE, AgreementProcess.AgreementState.SUSPENDED),
+          limit = 50,
+          offset = offset
+        )
+        .map(_.results.toList)
+
+    def go(start: Int)(as: List[AgreementProcess.Agreement]): Future[List[AgreementProcess.Agreement]] =
+      getAgreementsFrom(start).flatMap(agrs =>
+        if (agrs.isEmpty) Future.successful(as) else go(start + agrs.size)(as ++ agrs)
```

Comment:
I really like this clean solution ❤️ 
The only point I could rise is that using the empty array as exit condition means executing 1 more request to the agreement process each time this endpoint is called (that will always return an empty array).
While using `totalCount` allows to stop exactly when all results have been received

Reply:
This impl
```suggestion
        if (agrs.size < 50) Future.successful(as ++ agrs) else go(start + 50)(as ++ agrs)
```
would avoid a further call too, right?

Reply:
It would make an additional call just when `totalCount % 50 == 0`, but your suggestion is a good enough optimization IMO 👍